File: java/1.4 Iteradores y Colecciones de Objetos/1.4.3 Colecciones/Ejemplo04/com/ditamex/controller/ClienteController.java
    package com.ditamex.controller;
    
    import com.ditamex.data.Cliente;
    import java.util.ArrayList;
    import java.util.Collections;
    import java.util.Comparator;
    
    public class ClienteController
{
    /*
     * Revisar la documentación de la interfaz Comparator
     * https://docs.oracle.com/javase/8/docs/api/java/util/Comparator.html
     */
    
    public ClienteController(){
        ArrayList<Cliente> clientes = new ArrayList<Cliente>();
        
        clientes.add(new Cliente(101345, "Juan Rosas Pérez", "juanrp@gmail.com"));
        clientes.add(new Cliente(101289, "Ma. de Lourdes López Santiago", "mlls@yahoo.com"));
        clientes.add(new Cliente(101821, "Roberto Gómez Juárez", "roberto135@empresa.com"));
        clientes.add(new Cliente(101676, "Ana Fuentes Granados", "afg@ditamex.com"));
        clientes.add(new Cliente(101123, "Lucia Murrieta Flores", "lmurrietaf@sutienda.com"));
        
        System.out.println("- Order original -");
        for (Cliente cliente: clientes) {
            System.out.println(cliente);
        }
        
        Collections.sort(clientes);

        System.out.println("- Orderados por No. de Cliente -");
        for (Cliente cliente: clientes) {
            System.out.println(cliente);
        }
        
        Comparator<Cliente> comparadorPorNombre = new Comparator<Cliente>(){
            @Override
            public int compare(Cliente cliente1, Cliente cliente2) {
                return ( cliente1.getNombre().compareTo( cliente2.getNombre() ) );
            }
        };
        
        Collections.sort(clientes, comparadorPorNombre);

        System.out.println("- Orderados por Nombre -");
        for (Cliente cliente: clientes) {
            System.out.println(cliente);
        }
        
        
    }
    
    public static void main(String args[]) {
        new ClienteController();
    }
    
}

File: c++/1.4 Iteradores y Colecciones de Objetos/1.4.3 Colecciones/Ejemplo04/ClienteController.cpp
#include <iostream>
#include <list>

#include "Cliente.h"

using namespace std;

void imprimeLista(string nombreLista, list<Cliente> miLista) {
  cout << nombreLista << ": \n";
  for (list<Cliente>::iterator it = miLista.begin(); it != miLista.end();) {
    cout << it->toString();
    if (++it != miLista.end()) {
      cout << "\n";
    }
  }
  cout << endl;
}

class ComparadorPorNombre { 
  //clase con la función para comparar clientes por nombre
  //opcionalmente se puede crear en un archivo separado .h
 public:
  bool operator()( Cliente &cliente1,  Cliente &cliente2) {
    return cliente1.getNombre() < cliente2.getNombre();
  }
};

int main(int argc, char const *argv[]) {
  list<Cliente> clientes;

  clientes.push_back(Cliente(101345, "Juan Rosas Pérez", "juanrp@gmail.com"));
  clientes.push_back(Cliente(101289, "Ma. de Lourdes López Santiago", "mlls@yahoo.com"));
  clientes.push_back(Cliente(101821, "Roberto Gómez Juárez", "roberto135@empresa.com"));
  clientes.push_back(Cliente(101676, "Ana Fuentes Granados", "afg@ditamex.com"));
  clientes.push_back(Cliente(101123, "Lucia Murrieta Flores", "lmurrietaf@sutienda.com"));

  imprimeLista("Clientes", clientes);

  clientes.sort(); //ordena por número de cliente
  cout << "- Orderados por No. de Cliente -" << endl;
  imprimeLista("Clientes", clientes);

  ComparadorPorNombre comparadorPorNombre;
  clientes.sort(comparadorPorNombre); //ordena por nombre de cliente
  cout << "- Orderados por Nombre -" << endl;
  imprimeLista("Clientes", clientes);

  return 0;
}
